docs/proposals/13-automated-seed-management.md

```diff
@@ -263,18 +283,66 @@ ManagedSeedSets are reconciled by a new *managed seed set controller* in `garden
+
+* Ordered, graceful deployment and scaling.
+* Ordered, automated rolling updates. Such rolling updates can be *partitioned* (limited to replicas with ordinal numbers greater than or equal to the "partition") to achieve *phased rollouts*.
+
+A ManagedSeedSet, unlike a StatefulSet, does not need to maintain a stable identity for its ManagedSeeds. Furthermore, it would not be practical to always remove the replicas with the largest ordinal numbers when scaling in, since the corresponding seeds may have shoots scheduled onto them, while other seeds, with lower ordinals, may have fewer shoots (or none), and therefore be much better candidates for being removed.
+
+On the other hand, it would be beneficial if a ManagedSeedSet, like a StatefulSet, provides ordered deployment and scaling, ordered rolling updates, and phased rollouts. The main advantage of these features is that a deployment or update failure would affect fewer replicas (ideally just one), containing any potential damage and making the situation easier to handle, thus achieving some of the goals stated in [Issue #87](https://github.com/gardener/gardener/issues/87).
+
+Based on the above considerations, we propose the following mechanism for handling ManagedSeed identity and order:
+
+* A ManagedSeedSet uses *ordinal numbers generated by an increasing sequence* to identify ManagedSeeds and Shoots it creates and manages. These numbers always starts from 0 and are incremented by 1 for each newly added replica. 
+* Replicas (both ManagedSeeds and Shoots) are named after the ManagedSeedSet with the ordinal number appended. For example, for a ManagedSeedSet named `test` its replicas are named `test-0`, `test-1`, etc.
+* Gaps in the sequence created by removing replicas with ordinal numbers in the middle of the range are never filled in. A newly added replica always receives a number that is not only free, but also unique to itself. For example, if there are 2 replicas named `test-0` and `test-1` and any one of them is removed, a newly added replica will still be named `test-2`.   
+
+Although such ordinal numbers cannot provide stable identity, they can provide a predictable ordering for deployments and updates, and can also be used to partition rolling updates similarly to StatefulSet ordinal numbers.
+
+#### Update Strategies
+
+The ManagedSeedSet's `.spec.updateStrategy` field allows configuring automated rolling updates for the ManagedSeeds and Shoots in a ManagedSeedSet.
+
+**Rolling Updates**
+
+The `RollingUpdate` update strategy implements automated, rolling update for the ManagedSeeds and Shoots in a ManagedSeedSet. With this strategy, the ManagedSeedSet controller will update each ManagedSeed and Shoot in the ManagedSeedSet. It will proceed from the largest number to the smallest, updating each ManagedSeed and its corresponding Shoot one at a time. It will wait until both the Shoot and the Seed of an updated ManagedSeed are Ready prior to updating its predecessor.
```

Comment:
What about the Shoots that are deployed on those ManagedSeeds? 
In a bad case, a ManagedSeed. (the Shoot and Seed) are READY, but the control plane pods have troubles. E.g there was an update to the kubelet configuration for a certain worker pool of the Seed. cluster.

The rollout would then only continue if not more than X percent of the Shoot clusters on the Seed have Reconcile or Readiness (only control plane) problems. 

WDYT

Reply:
OK, makes sense in general, not sure how complex that would be to implement. Perhaps before implementing the ManagedSeedSet controller, we should consider introducing a "managed seed care" controller that would update the ManagedSeed resource itself with the health of its Seed and Shoots deployed on top, since keeping track of all these might be too much for the ManagedSeedSet controller to handle.

Reply:
Ok, I like that idea. Also makes the initial implementation easier

Reply:
OK, I will mention this in the document.